app_dart/lib/src/request_handlers/refresh_chromebot_status.dart

```diff
@@ -79,19 +80,82 @@ class RefreshChromebotStatus extends ApiRequestHandler<Body> {
       DatastoreService datastore,
       Map<LuciBuilder, List<LuciTask>> luciTasks) async {
     for (String branch in branches) {
-      await for (FullTask task in datastore.queryRecentTasks(
-          taskName: builder.taskName, branch: branch)) {
-        for (LuciTask luciTask in luciTasks[builder]) {
-          if (luciTask.commitSha == task.commit.sha &&
-              luciTask.ref == 'refs/heads/$branch') {
+      final List<FullTask> tasks = await datastore
+          .queryRecentTasks(taskName: builder.taskName, branch: branch)
+          .toList();
+      final List<LuciTask> allLuciTasks = luciTasks[builder];
+      final Set<LuciTask> updatedLuciTasks = <LuciTask>{};
+
+      /// Scan both [tasks] and [luciTasks] from old to new, since [luciTasks] may
+      /// contain new re-run builds which are not in datastore yet. If matched, then
+      /// update accordingly.
+      for (int i = tasks.length - 1; i >= 0; i--) {
+        final FullTask task = tasks[i];
+        for (int j = allLuciTasks.length - 1; j >= 0; j--) {
+          final LuciTask luciTask = allLuciTasks[j];
+          if (_taskMatched(task, luciTask, branch) &&
+              !updatedLuciTasks.contains(luciTask)) {
             final Task update = task.task;
             update.status = luciTask.status;
+            update.buildId = luciTask.buildId;
             await datastore.insert(<Task>[update]);
-            // Stop updating task whenever we find the latest status of the same commit.
+            updatedLuciTasks.add(luciTask);
             break;
           }
         }
       }
+
+      final List<Task> newTasks =
+          _getNewTasks(updatedLuciTasks, builder, allLuciTasks, datastore);
```

Comment:
Why we need to filter again with a new loop? can we filter instead in previous iteration?

Reply:
To avoid confusion, I decide to skip the `newTasks` insert in this PR. Instead, will use https://github.com/flutter/flutter/issues/58268 to take care of new re-run builds.